text/0000-block-manager.md

```diff
@@ -0,0 +1,367 @@
+and interfaces. The requirements for the design presented in this RFC are that
+it:
+
+- Shall store blocks for later
+- Shall ensure integrity of blocks stored
+- Shall provide a mechanism for preventing required blocks from being dropped
+- Shall provide a mechanism for persisting to a store
+
+The constraints for the design presented in this RFC are that:
+
+- The API must work for both forking vs non-forking consensus mechanisms
+- The API defined must be reusable, meaning it provides generic methods that
+  make few assumptions about how the API will be used
+
+The following new concepts are used by this design and are defined here:
+
+For a collection of blocks, __integrity__ is defined to mean that every block
+in the collection is accompanied by its parent.
+
+A __branch__ is a collection of blocks such that:
+
+1. The blocks are sorted in order of increasing block number
+2. The branch has integrity
+
+A __block tree__ is a collection of branches with a single common ancestor or
+__root__.
+
+The following is a visualization of an example block tree containing blocks
+{0, A, B, C, ..., H}.
+
+```
+0 (root)
+|
+|\
+A |
+| E
+|\
+B |
+| F
+| |\
+C | |
+| G |
+|   H
+D
+```
+
+In the above example, the block tree contains multiple branches. They are (0,
+A, B, C, D), (0, E), (0, A, F, G), and (0, A, F, H).
+
+In order to satisfy the requirements and constraints of this RFC, the
+__block manager__ class and __block store__ interface were created.
+
+A __block store__ is an interface which supports storing and retrieving blocks.
+It can be backed by any type of storage, for example disk or network. Because a
+blockchain's length grows indefinitely, part of it eventually has to be
+archived in some way. The __block store__ interface is intended to help solve
+the problem of persisting parts of the __block tree__ in a way that does not
+make assumptions about how it is persisted or what parts of the tree are
+persisted.
+
+The __block manager__ is a collection of __block tree__ s (though it usually
+contains 1) that also supports methods for:
+
+- Ensuring that blocks depended on by external objects are not removed.
+- Persisting blocks to a __block store__.
+
+In addition to the __block manager__ class and __block store__ interface, a
+__commit store__ class was created The commit store is a __block store__
+implementation that can be added to the block tree for persisting committed
+blocks to disk. It also implements an LRU cache which holds some number of most
+recently accessed blocks in memory. The __commit store__ replaces the what was
+previously called the "block store".
+
+## Block Manager API
+
+The following is a listing of the __block manager__ methods and their function:
+
+`put(branch: List<Block>)`
+
+Atomically adds the given blocks to the tree.
+
+The blocks passed must form a branch meaning they satisfy the following
+conditions:
+
+1. The first block's parent is present
```

Comment:
Does present mean that the parent block is already in the block manager? It is ambiguous as to whether is is that or "present in the List<Block>".

Reply:
I read that to mean that the parent of the list[0] element is already present in tree managed by the block manager.